### PR TITLE
Implement static_atom_encoder option

### PR DESCRIPTION
**Pull request by arjan**
This PR introduces the `static_atom_encoder` option to `Code.string_to_quoted/2`, as described in [this elixir-lang-core thread](https://groups.google.com/forum/#!topic/elixir-lang-core/1A0kOswSB7U).

I think it is ready for review; code, docs and 2 added tests.

### CODE REVIEW

File: lib/elixir/src/elixir_tokenizer.erl [hunk 1]
```diff
@@ -807,6 +809,16 @@ is_unnecessary_quote([Part], #elixir_tokenizer{warn_on_unnecessary_quotes=true}
 is_unnecessary_quote(_Parts, _Scope) ->
   false.
 
+unsafe_to_atom(Part, Line, Column, #elixir_tokenizer{static_atoms_encoder=StaticAtomsEncoder} = Scope) when
+    is_function(StaticAtomsEncoder) ->
```
Comment (whatyouhide) [inline lib/elixir/src/elixir_tokenizer.erl]: We can specify the arity of the `StaticAtomsEncoder` function:

```suggestion
    is_function(StaticAtomsEncoder, 2) ->
```

File: lib/elixir/test/elixir/code_test.exs [hunk 2]
```diff
@@ -153,6 +153,36 @@ defmodule CodeTest do
                {:error, {1, "unsafe atom does not exist: ", "there_is_no_such_atom"}}
     end
 
+    test "supports static_atoms_encoder" do
+      ref = make_ref()
+
+      encoder = fn "there_is_no_such_atom", metadata ->
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/code_test.exs]: Nitpick, but maybe for better assertion errors we can have:

```elixir
encoder = fn atom, meta ->
  assert atom == "ther_is_no_such_atom"
  assert meta[:line] == =1
  # ...same
end
```

File: lib/elixir/src/elixir_tokenizer.erl [hunk 3]
```diff
@@ -807,6 +809,16 @@ is_unnecessary_quote([Part], #elixir_tokenizer{warn_on_unnecessary_quotes=true}
...
+  Metadata = [{line, Line}, {column, Column}, {file, Scope#elixir_tokenizer.file}],
+  Value = elixir_utils:characters_to_binary(Part),
+  case StaticAtomsEncoder(Value, Metadata) of
+    {ok, Term} ->
+      {ok, Term};
+    {error, Reason} ->
+      {error, {Line, Column, Reason, Part}}
+  end;
```
Comment (whatyouhide) [inline lib/elixir/src/elixir_tokenizer.erl]: Does having this clause before the next one mean that `:statis_atoms_encoder` will let atoms that are bigger than 255 codepoints work? Since the tokenizer now raises an error for atoms that go over the system limit, maybe it's better to keep that behaviour?

File: lib/elixir/src/elixir_tokenizer.erl (hunk 1 above)
Comment (josevalim) [reply lib/elixir/src/elixir_tokenizer.erl]: Since this is called during the operation and not during input validation, I think it should stay as is, otherwise you can pass a function with arity 1 and never know why it was not called.

File: lib/elixir/test/elixir/code_test.exs [hunk 4]
```diff
@@ -153,6 +153,36 @@ defmodule CodeTest do
...
+          ArgumentError ->
+            {:ok, {:user_atom, string}}
+        end
+      end
+
+      assert {:ok, {:try, _, [[do: {:test, _, [{{:user_atom, "atom_does_not_exist"}, _, []}]}]]}} =
+               Code.string_to_quoted("try do: test(atom_does_not_exist())",
+                 static_atoms_encoder: encoder
+               )
+    end
```
Comment (josevalim) [inline lib/elixir/test/elixir/code_test.exs]: We also need a test for the `{:error, reason}` branch. For instance, I think we will have to convert the Reason to a charlist and the current code does not do it!

File: lib/elixir/src/elixir_tokenizer.erl (hunk 3 above)
Comment (arjan) [reply lib/elixir/src/elixir_tokenizer.erl]: I actually did this on purpose because when you pass an atom encoder, I think it should be up to yourself to check limits like these. I can imagine an encoder function that takes too long atom names and transforms them into atoms that fit the system limit.

File: lib/elixir/test/elixir/code_test.exs (hunk 4 above)
Comment (arjan) [reply lib/elixir/test/elixir/code_test.exs]: Yep you are right. Sorry I forgot, will add this. The return value for the error was not entirely clear for me (the spec just says `{:error, {line :: pos_integer, term, term}}`).

File: lib/elixir/test/elixir/code_test.exs (hunk 2 above)
Comment (arjan) [reply lib/elixir/test/elixir/code_test.exs]: Addressed this in 23e3b95

File: lib/elixir/test/elixir/code_test.exs (hunk 4 above)
Comment (arjan) [reply lib/elixir/test/elixir/code_test.exs]: Fix this in 282675e0570a1e0f6e7be0579f7495e11ba4bf9d

Note that it adds `": "` to the error message. Which is kind of weird but it is in line with the other error messages. Alternatively it could not do that but then the encoder callback function is supposed to add it, which is also questionable..

File: lib/elixir/src/elixir_tokenizer.erl (hunk 3 above)
Comment (whatyouhide) [reply lib/elixir/src/elixir_tokenizer.erl]: @arjan I am still not sure about this. I think it depends on what we want this functionality to address. If we want this to let people parse random code safely, then I would go with keeping the hard limit on 255 characters.

File: lib/elixir/lib/code.ex [hunk 5]
```diff
@@ -650,6 +650,10 @@ defmodule Code do
       when non-existing atoms are found by the tokenizer.
       Defaults to `false`.
 
+    * `:static_atom_encoder` - The static atom encoder function, see
+      the section below. The `:static_atom_encoder` option overrides
```
Comment (whatyouhide) [inline lib/elixir/lib/code.ex]: Try to avoid pointing to just "the section below" since this quickly becomes outdated if we move documentation around. Use this instead:

```suggestion
      the ["The `:static_atom_encoder` function" section](module-the-static_atom_encoder-function)
      below. The `:static_atom_encoder` option overrides
```

File: lib/elixir/lib/code.ex [hunk 6]
```diff
@@ -659,6 +663,33 @@ defmodule Code do
   The opposite of converting a string to its quoted form is
   `Macro.to_string/2`, which converts a quoted form to a string/binary
   representation.
+
+  ## The `:static_atom_encoder` function
+
+  When `static_atom_encoder: &my_encoder/2` is passed as an argument,
+  `my_encoder/2` is called every time the tokenizer needs to create a
+  'static' atom. Static atoms are atoms in the AST that function as
```
Comment (whatyouhide) [inline lib/elixir/lib/code.ex]: Nitpick, but we don't use single quotes `'` anywhere in the Elixir docs that I know of.

```suggestion
  "static" atom. Static atoms are atoms in the AST that function as
```

File: lib/elixir/lib/code.ex [hunk 7]
```diff
@@ -659,6 +663,33 @@ defmodule Code do
...
+  however, in that case the AST is no longer "valid" in that it cannot
+  be used to compile or evaluate Elixir code. A use case for this is
+  if you want to use the Elixir parser in a user-facing situation, but
+  you don't want to exhaust the atom table.
+
+  When the encoder fails for some reason, it must return `{:error,
+  reason :: binary}`. The atom encoder is not called for *all* atoms
+  that are present in the AST: it won't be invoked for operators,
+  syntax keywords (fn, do, etc), and neither for atoms containing
+  interpolations.
```
Comment (whatyouhide) [inline lib/elixir/lib/code.ex]: This is an important note, so I would write it in a more formal way:

```
The atom encoder is not called for *all* atoms that are present in
the AST. It won't be invoked for the following atoms:

  * operators (`:+`, `:-`, and so on)

  * syntax keywords (`fn`, `do`, and so on)

  * atoms containing interpolation (`:"#{1 + 1} is two"`)
```

Wdyt?

Comment (arjan) [reply lib/elixir/lib/code.ex]: Yep, looks good, done in f9ae8e7

File: lib/elixir/src/elixir_tokenizer.erl (hunk 3 above)
Comment (arjan) [reply lib/elixir/src/elixir_tokenizer.erl]: I'm fine with changing this behaviour around so that any atoms-to-be are first length-checked. It's an edge case anyway.

Comment (josevalim) [reply lib/elixir/src/elixir_tokenizer.erl]: I think 255 is sane limit to have anyway, so let's move the limit check clause up.

File: lib/elixir/test/elixir/code_test.exs (hunk 4 above)
Comment (josevalim) [reply lib/elixir/test/elixir/code_test.exs]: :+1:

File: lib/elixir/lib/code.ex [hunk 8]
```diff
@@ -650,6 +650,10 @@ defmodule Code do
       when non-existing atoms are found by the tokenizer.
       Defaults to `false`.
 
+    * `:static_atom_encoder` - The static atom encoder function, see
+      "The `:static_atom_encoder` function" section below. This option
+      overrides the `:existing_atoms_only` option.
```
Comment (josevalim) [inline lib/elixir/lib/code.ex]: I removed the link because it doesn't work nicely when read on the terminal. :(

File: lib/elixir/src/elixir_tokenizer.erl (hunk 1 above)
Comment (umamaistempo) [reply lib/elixir/src/elixir_tokenizer.erl]: 🤔  IMO, such case should be treated as a different clause but that would not be terse and would be a bit too defensive 😅

Comment (whatyouhide) [reply lib/elixir/src/elixir_tokenizer.erl]: Oh good catch @josevalim, this would only make the clause fail and cascade. Agreed, better to have the actual call fail with wrong number of arguments then.

**Comment by umamaistempo** [comment]
Adding that 255 check (and potentially pointing that limitation on the doc) and LGTM :+1:

File: lib/elixir/lib/code.ex (hunk 8 above)
Comment (umamaistempo) [reply lib/elixir/lib/code.ex]: :(

File: lib/elixir/src/elixir_tokenizer.erl (hunk 3 above)
Comment (arjan) [reply lib/elixir/src/elixir_tokenizer.erl]: Fixed in 970ba29

**Comment by whatyouhide** [comment]
Great job, thanks @arjan! 💟

**Comment by arjan** [comment]
Thanks for merging! 😁
